In [2]:
import numpy as np
import pandas as pd

# Задание 1
Изучить представленный набор данных на основе описания его столбцов, загрузить его и оставить 8 столбцов для дальнейшего изучения: surgery?, Age, rectal temperature, pulse, respiratory rate, temperature of extremities, pain, outcome.

In [21]:
data = pd.read_csv('horse_data.csv', 
                    names = ['surgery?', 
                             'age', 
                             'hospital_num', 
                             'rectal_temperature', 
                             'pulse', 
                             'respiratory_rate',
                             'temp_of_extremities',
                             'peripheral pulse',
                             'mucous membranes',                
                             'capillary refill time', 
                             'pain', 
                             'peristalsis', 
                             'abdominal_distension', 
                             'nasogastric tube', 
                             'nasogastric reflux', 
                             'nasogastric reflux PH', 
                             'rectal examination - feces', 
                             'abdomen', 
                             'packed cell volume', 
                             'total protein', 
                             'abdominocentesis appearance',
                             'abdomcentesis total protein',
                             'outcome',
                             'surgical lesion?',
                             '25', 
                             '26',
                             '27',
                             'cp_data'], 
                   na_values = '?')
horse = data[['surgery?', 'age', 'rectal_temperature', 'pulse', 'respiratory_rate', 'temp_of_extremities', 'pain', 'outcome']]
horse.head(50)

surgery?  age  rectal_temperature  pulse  respiratory_rate  \
0        2.0    1                38.5   66.0              28.0   
1        1.0    1                39.2   88.0              20.0   
2        2.0    1                38.3   40.0              24.0   
3        1.0    9                39.1  164.0              84.0   
4        2.0    1                37.3  104.0              35.0   
5        2.0    1                 NaN    NaN               NaN   
6        1.0    1                37.9   48.0              16.0   
7        1.0    1                 NaN   60.0               NaN   
8        2.0    1                 NaN   80.0              36.0   
9        2.0    9                38.3   90.0               NaN   
10       1.0    1                38.1   66.0              12.0   
11       2.0    1                39.1   72.0              52.0   
12       1.0    1                37.2   42.0              12.0   
13       2.0    9                38.0   92.0              28.0   
14       1.0    1                38.2   76.0              28.0   
15       1.0    1                37.6   96.0              48.0   
16       1.0    9                 NaN  128.0              36.0   
17       2.0    1                37.5   48.0              24.0   
18       1.0    1                37.6   64.0              21.0   
19       2.0    1                39.4  110.0              35.0   
20       1.0    1                39.9   72.0              60.0   
21       2.0    1                38.4   48.0              16.0   
22       1.0    1                38.6   42.0              34.0   
23       1.0    9                38.3  130.0              60.0   
24       1.0    1                38.1   60.0              12.0   
25       2.0    1                37.8   60.0              42.0   
26       1.0    1                38.3   72.0              30.0   
27       1.0    1                37.8   48.0              12.0   
28       1.0    1                 NaN    NaN               NaN   
29       2.0    1                37.7   48.0               NaN   
30       2.0    1                37.7   96.0              30.0   
31       2.0    1                37.2  108.0              12.0   
32       1.0    1                37.2   60.0               NaN   
33       1.0    1                38.2   64.0              28.0   
34       1.0    1                 NaN  100.0              30.0   
35       2.0    1                 NaN  104.0              24.0   
36       2.0    1                38.3  112.0              16.0   
37       1.0    1                37.8   72.0               NaN   
38       2.0    1                38.6   52.0               NaN   
39       1.0    9                39.2  146.0              96.0   
40       1.0    1                 NaN   88.0               NaN   
41       2.0    9                39.0  150.0              72.0   
42       2.0    1                38.0   60.0              12.0   
43       1.0    1                 NaN  120.0               NaN   
44       1.0    1                35.4  140.0              24.0   
45       2.0    1                 NaN  120.0               NaN   
46       1.0    1                37.9   60.0              15.0   
47       2.0    1                37.5   48.0              16.0   
48       1.0    1                38.9   80.0              44.0   
49       2.0    1                37.2   84.0              48.0   

    temp_of_extremities  pain  outcome  
0                   3.0   5.0      2.0  
1                   NaN   3.0      3.0  
2                   1.0   3.0      1.0  
3                   4.0   2.0      2.0  
4                   NaN   NaN      2.0  
5                   2.0   2.0      1.0  
6                   1.0   3.0      1.0  
7                   3.0   NaN      2.0  
8                   3.0   4.0      3.0  
9                   1.0   5.0      1.0  
10                  3.0   3.0      1.0  
11                  2.0   2.0      1.0  
12                  2.0   3.0      1.0  
13                  1.0   1.0      2.0  
14                  3.

# Задание 2
Проанализировать значения по столбцам, рассчитать базовые статистики, найти выбросы.

In [30]:
# age
horse.age.mode()[0]
horse.age.value_counts()

1    300
Name: age, dtype: int64

Самые частые лошади в выборке - "1", т.е. "Adult horse"
Значения "9" скорее всего выбросы

In [29]:
horse.age.replace(9, 1, inplace=True)

In [23]:
# rectal temperature
print(f'Максимальный: {max(horse.rectal_temperature)}')
print(f'Минимальный: {min(horse.rectal_temperature)}')
print(f'Размах: {max(horse.rectal_temperature) - min(horse.rectal_temperature)}')
print(f'Ср.арифм.: {horse.rectal_temperature.mean()}')
print(f'Мода: {horse.rectal_temperature.mode()[0]}')
print(f'Медиана: {horse.rectal_temperature.median()}')
print(f'СКО: {horse.rectal_temperature.std()}')
print(f'Межквартильный размах: {horse.rectal_temperature.quantile(0.75) - horse.rectal_temperature.quantile(0.25)}')


Максимальный: 40.8
Минимальный: 35.4
Размах: 5.399999999999999
Ср.арифм.: 38.16791666666669
Мода: 38.0
Медиана: 38.2
СКО: 0.7322886641121578
Межквартильный размах: 0.7000000000000028


Чаще всего в выборку попадали лошади с температурой "38". Не уверено, но возможно это нормальная для них темпаратура? 

In [24]:
# pulse
print(f'Максимальный: {max(horse.respiratory_rate)}')
print(f'Минимальный: {min(horse.respiratory_rate)}')
print(f'Размах: {max(horse.respiratory_rate) - min(horse.respiratory_rate)}')
print(f'Ср.арифм.: {horse.respiratory_rate.mean()}')
print(f'Мода: {horse.respiratory_rate.mode()[0]}')
print(f'Медиана: {horse.respiratory_rate.median()}')
print(f'СКО: {horse.respiratory_rate.std()}')
print(f'Межквартильный размах: {horse.respiratory_rate.quantile(0.75) - horse.respiratory_rate.quantile(0.25)}')

Максимальный: 184.0
Минимальный: 30.0
Размах: 154.0
Ср.арифм.: 71.91304347826087
Мода: 48.0
Медиана: 64.0
СКО: 28.630556660735003
Межквартильный размах: 40.0


С пульсом немного непонятно, модальное значение "48", т.е. скорее всего умирающие лошади, поэтому есть вероятность того, что в статистику чаще всего попадают лошади на поздних этапах болезни
Максимальные значения "184" вероятнее всего выбросы

In [26]:
# respiratory rate
print(f'Максимальный: {max(horse.respiratory_rate)}')
print(f'Минимальный: {min(horse.respiratory_rate)}')
print(f'Размах: {max(horse.respiratory_rate) - min(horse.respiratory_rate)}')
print(f'Ср.арифм.: {horse.respiratory_rate.mean()}')
print(f'Мода: {horse.respiratory_rate.mode()[0]}')
print(f'Медиана: {horse.respiratory_rate.median()}')
print(f'СКО: {horse.respiratory_rate.std()}')
print(f'Межквартильный размах: {horse.respiratory_rate.quantile(0.75) - horse.respiratory_rate.quantile(0.25)}')

Максимальный: 96.0
Минимальный: 8.0
Размах: 88.0
Ср.арифм.: 30.417355371900825
Мода: 20.0
Медиана: 24.5
СКО: 17.642231385134664
Межквартильный размах: 17.5


Максимальные значения "96" - вероятнее всего выбросы. Нельзя же так часто дышать

In [27]:
# temp_of_extremities
print(f'Максимальный: {max(horse.temp_of_extremities)}')
print(f'Минимальный: {min(horse.temp_of_extremities)}')
print(f'Размах: {max(horse.temp_of_extremities) - min(horse.temp_of_extremities)}')
print(f'Ср.арифм.: {horse.temp_of_extremities.mean()}')
print(f'Мода: {horse.temp_of_extremities.mode()[0]}')
print(f'Медиана: {horse.temp_of_extremities.median()}')
print(f'СКО: {horse.temp_of_extremities.std()}')
print(f'Межквартильный размах: {horse.temp_of_extremities.quantile(0.75) - horse.temp_of_extremities.quantile(0.25)}')

Максимальный: 4.0
Минимальный: 1.0
Размах: 3.0
Ср.арифм.: 2.348360655737705
Мода: 3.0
Медиана: 3.0
СКО: 1.0450536920112758
Межквартильный размах: 2.0


In [31]:
# pain
print(f'Максимальный: {max(horse.pain)}')
print(f'Минимальный: {min(horse.pain)}')
print(f'Размах: {max(horse.pain) - min(horse.pain)}')
print(f'Ср.арифм.: {horse.pain.mean()}')
print(f'Мода: {horse.pain.mode()[0]}')
print(f'Медиана: {horse.pain.median()}')
print(f'СКО: {horse.pain.std()}')
print(f'Межквартильный размах: {horse.pain.quantile(0.75) - horse.pain.quantile(0.25)}')

Максимальный: 5.0
Минимальный: 1.0
Размах: 4.0
Ср.арифм.: 2.951020408163265
Мода: 3.0
Медиана: 3.0
СКО: 1.3079395711243502
Межквартильный размах: 2.0


In [32]:
# outcome
print(f'Максимальный: {max(horse.outcome)}')
print(f'Минимальный: {min(horse.outcome)}')
print(f'Размах: {max(horse.outcome) - min(horse.outcome)}')
print(f'Ср.арифм.: {horse.outcome.mean()}')
print(f'Мода: {horse.outcome.mode()[0]}')
print(f'Медиана: {horse.outcome.median()}')
print(f'СКО: {horse.outcome.std()}')
print(f'Межквартильный размах: {horse.outcome.quantile(0.75) - horse.outcome.quantile(0.25)}')

Максимальный: 3.0
Минимальный: 1.0
Размах: 2.0
Ср.арифм.: 1.5518394648829432
Мода: 1.0
Медиана: 1.0
СКО: 0.7371869448054933
Межквартильный размах: 1.0


# Задание 3. Работа с пропусками

In [43]:
for col in horse.columns:
    pct_missing = horse[col].isnull().mean()
    print(f'{col} - {pct_missing :.1%}')

surgery? - 0.3%
age - 0.0%
rectal_temperature - 0.0%
pulse - 0.0%
respiratory_rate - 0.0%
temp_of_extremities - 0.0%
pain - 0.0%
outcome - 0.3%


1. Можно ничего не менять, т.к. количество пропусков незначительное
2. Пропуски отсутствуют
3. Можно заменить модой, т.к. это наиболее часто встречаемое значение в выборке
4. Пульс заполнить медианой, поскольку пропусков по пульсу не много, но они значимые
5. Аналогично с п.4
6. temp_of_extremities - заполнить модой
7. Аналогично с п.4

In [35]:

horse.rectal_temperature.fillna(horse.rectal_temperature.mode()[0], inplace=True)

In [37]:

horse.pulse.fillna(horse.pulse.median(), inplace=True)

In [39]:
horse.respiratory_rate.fillna(horse.respiratory_rate.median(), inplace=True)

In [41]:
horse.temp_of_extremities.fillna(horse.temp_of_extremities.mode()[0], inplace=True)

In [42]:
horse.pain.fillna(horse.pain.median(), inplace=True)